File: services/graphql-dms/src/models/delivery_routes.rs
use std::sync::Arc;

use async_graphql::{ComplexObject, Context, dataloader::Loader};
use chrono::{DateTime, NaiveDate, Utc};
use graphql_core::PostgresDataLoader;
use graphql_tms::models::drivers;
use sqlx::prelude::FromRow;
use uuid::Uuid;

use super::enums::DeliveryRouteStatusEnum;

#[derive(Debug, Clone, Copy, PartialEq, Hash, Eq)]
pub struct PrimaryKey(pub Uuid);

#[derive(Clone, Debug, PartialEq, async_graphql :: SimpleObject, FromRow)]
#[graphql(name = "DmsDeliveryRoutes", complex)]
pub struct Model {
    pub id: Uuid,
    #[graphql(skip)]
    pub driver_id: Uuid,
    pub route_date: NaiveDate,
    pub status: Option<DeliveryRouteStatusEnum>,
    pub optimized_route_data: Option<String>,
    pub total_distance_km: Option<f32>,
    pub estimated_duration_minutes: Option<i32>,
    pub actual_duration_minutes: Option<i32>,
    pub started_at: Option<DateTime<Utc>>,
    pub completed_at: Option<DateTime<Utc>>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
}

#[ComplexObject]
impl Model {
    async fn driver(&self, ctx: &Context<'_>) -> async_graphql::Result<drivers::Model> {
        let loader = ctx.data::<async_graphql::dataloader::DataLoader<PostgresDataLoader>>()?;

        Ok(loader
            .load_one(drivers::PrimaryKey(self.driver_id))
            .await?
            .ok_or(async_graphql::Error::new("Unable to find driver"))?)
    }
}

impl Loader<PrimaryKey> for PostgresDataLoader {
    type Error = Arc<sqlx::Error>;
    type Value = Model;

    async fn load(
        &self,
        keys: &[PrimaryKey],
    ) -> Result<std::collections::HashMap<PrimaryKey, Self::Value>, Self::Error> {
        let keys = keys.iter().map(|k| k.0).collect::<Vec<_>>();

        let results = sqlx::query_as::<_, Self::Value>(
            "select * from dms.delivery_routes where id = ANY($1)",
        )
        .bind(&keys)
        .fetch_all(&self.pool)
        .await?
        .into_iter()
        .map(|model| (PrimaryKey(model.id), model))
        .collect::<_>();

        Ok(results)
    }
}

File: src/graphql/queries/dms/deliveryRoutes.ts
import { graphql } from "@/lib/graphql/client";

// ============================================================================
// DELIVERY ROUTE OPERATIONS
// ============================================================================

/**
 * Fetches a single delivery route by its ID.
 * @param id The UUID of the delivery route.
 */
export const getDeliveryRoute = graphql(`
  query GetDeliveryRoute($id: UUID!) {
    dms {
      deliveryRoute(id: $id) {
        id
        routeDate
        status
        optimizedRouteData
        totalDistanceKm
        estimatedDurationMinutes
        actualDurationMinutes
        startedAt
        completedAt
        createdAt
        updatedAt
        driver {
          id
          licenseNumber
        }
      }
    }
  }
`);

/**
 * Fetches a paginated list of delivery routes.
 * @param limit The maximum number of items to return.
 * @param page The page number to retrieve.
 */
export const getDeliveryRoutes = graphql(`
  query GetDeliveryRoutes($limit: Int!, $page: Int!) {
    dms {
      deliveryRoutes(limit: $limit, page: $page) {
        id
        routeDate
        status
        optimizedRouteData
        totalDistanceKm
        estimatedDurationMinutes
        actualDurationMinutes
        startedAt
        completedAt
        createdAt
        updatedAt
        driver {
          id
          licenseNumber
        }
      }
    }
  }
`);
